File: Components/ASP.NET/CS/MVCxSpreadsheetDemos/Code/Helpers/SpreadsheetDemoUtils.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web;
using DevExpress.Web;
using DevExpress.Web.ASPxSpreadsheet;

namespace DevExpress.Web.Demos {
    public static class SpreadsheetDemoUtils {
        public static void HideFileTab(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet) {
            spreadsheet.CreateDefaultRibbonTabs(true);
            RemoveRibbonTab(spreadsheet, typeof(SRFileTab));
        }

        public static void CreateOverviewRibbon(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet) {
            spreadsheet.CreateDefaultRibbonTabs(true);
            RemoveRibbonTab(spreadsheet, typeof(SRFileTab));
            RemoveRibbonTab(spreadsheet, typeof(SRPageLayoutTab));
        }

        static void RemoveRibbonTab(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet, Type tabTypeToRemove) {
            foreach(RibbonTab tab in spreadsheet.RibbonTabs) {                
                if(tab.GetType() == tabTypeToRemove) {
                    spreadsheet.RibbonTabs.Remove(tab);
                   break;
                }
            }
        }

        public static void HideAllTabsExceptFileAndPageLayout(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet) {
            spreadsheet.CreateDefaultRibbonTabs(true);

            ShowRibbonItemsOnly(spreadsheet,  
                new Type[] { 
                    typeof(SRFilePrintCommand), 
                    typeof(SRPageSetupMarginsCommand), 
                    typeof(SRPageSetupOrientationCommand), 
                    typeof(SRPageSetupPaperKindCommand), 
                    typeof(SRPrintGridlinesCommand),
                    typeof(SRPrintHeadingsCommand)
                }
            );
        }

        public static void HideAllTabsExceptHomeAndView(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet) {
            spreadsheet.CreateDefaultRibbonTabs(true);

            RemoveRibbonTab(spreadsheet, typeof(SRFileTab));
            RemoveRibbonTab(spreadsheet, typeof(SRInsertTab));
            RemoveRibbonTab(spreadsheet, typeof(SRPageLayoutTab));
            RemoveRibbonTab(spreadsheet, typeof(SRPageLayoutTab));
            RemoveRibbonTab(spreadsheet, typeof(SRFormulasTab));
            RemoveRibbonTab(spreadsheet, typeof(SRDataTab));
        }

        public static void HidePageLayoutTab(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet) {
            RemoveRibbonTab(spreadsheet, typeof(SRPageLayoutTab));
        }

        public static void HideFormulasTab(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet) {
            RemoveRibbonTab(spreadsheet, typeof(SRFormulasTab));
        }

        private static void HideRibbonItems(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet, Type[] itemTypes) {
            ProcessRibbonItems(spreadsheet, itemTypes, false);
        }
        private static void ShowRibbonItemsOnly(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet, Type[] itemTypes) {
            ProcessRibbonItems(spreadsheet, itemTypes, true);
        }
        private static void ProcessRibbonItems(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet, Type[] targetTypes, bool removeNotInList) {
            var groups = new List<RibbonGroup>();
            var items = new List<RibbonItemBase>();
            
            List<RibbonTab> tabs = spreadsheet.RibbonTabs.ToList();
            foreach(RibbonTab tab in tabs) {                
                groups.AddRange(tab.Groups);
            }
            foreach(RibbonGroup group in groups) {
                items.AddRange(group.Items);
            }

            RemoveItemsByTypes(items, targetTypes, removeNotInList);
            RemoveEmptyGroups(groups);
            RemoveEmptyTabs(spreadsheet, tabs);
        }

        private static void RemoveItemsByTypes(List<RibbonItemBase> items, Type[] targetTypes, bool removeNotInList) {
            var targetItem = new List<RibbonItemBase>();
            var itemsNotInList = new List<RibbonItemBase>();
            GroupItemsByType(items, targetTypes, targetItem, itemsNotInList);

            if(removeNotInList)
                RemoveItems(itemsNotInList);
            else
                RemoveItems(targetItem);
                
        }

        private static void GroupItemsByType(List<RibbonItemBase> items, Type[] targetTypes, List<RibbonItemBase> targetItem, List<RibbonItemBase> itemsNotInList) {
            foreach(RibbonItemBase item in items) {
                if(targetTypes.Contains(item.GetType()))
                    targetItem.Add(item);
                else
                    itemsNotInList.Add(item);
            }
        }

        private static void RemoveItems(List<RibbonItemBase> items) {
            foreach(RibbonItemBase item in items)
                item.Group.Items.Remove(item);
        }
        private static void RemoveEmptyGroups(List<RibbonGroup> groups) {
            foreach(RibbonGroup group in groups) {
                if(group.Items.IsEmpty)
                    group.Tab.Groups.Remove(group);
            }
        }
        private static void RemoveEmptyTabs(ASPxSpreadsheet.ASPxSpreadsheet spreadsheet, List<RibbonTab> tabs) {
            foreach(RibbonTab tab in tabs) {
                if(tab.Groups.IsEmpty)
                    spreadsheet.RibbonTabs.Remove(tab); 
            }
        }
    }
}

File: Components/ASP.NET/VB/ASPxSpreadsheetDemos/App_Code/DemoHelpers/SpreadsheetDemoUtils.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Web
Imports DevExpress.Web
Imports DevExpress.Web.ASPxSpreadsheet

Namespace DevExpress.Web.Demos
	Public NotInheritable Class SpreadsheetDemoUtils
		Private Sub New()
		End Sub
		Public Shared Sub HideFileTab(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet)
			spreadsheet.CreateDefaultRibbonTabs(True)
			RemoveRibbonTab(spreadsheet, GetType(SRFileTab))
		End Sub

		Public Shared Sub CreateOverviewRibbon(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet)
			spreadsheet.CreateDefaultRibbonTabs(True)
			RemoveRibbonTab(spreadsheet, GetType(SRFileTab))
			RemoveRibbonTab(spreadsheet, GetType(SRPageLayoutTab))
		End Sub

		Private Shared Sub RemoveRibbonTab(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet, ByVal tabTypeToRemove As Type)
			For Each tab As RibbonTab In spreadsheet.RibbonTabs
				If tab.GetType() Is tabTypeToRemove Then
					spreadsheet.RibbonTabs.Remove(tab)
				   Exit For
				End If
			Next tab
		End Sub

		Public Shared Sub HideAllTabsExceptFileAndPageLayout(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet)
			spreadsheet.CreateDefaultRibbonTabs(True)

			ShowRibbonItemsOnly(spreadsheet, New Type() { GetType(SRFilePrintCommand), GetType(SRPageSetupMarginsCommand), GetType(SRPageSetupOrientationCommand), GetType(SRPageSetupPaperKindCommand), GetType(SRPrintGridlinesCommand), GetType(SRPrintHeadingsCommand) })
		End Sub

		Public Shared Sub HideAllTabsExceptHomeAndView(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet)
			spreadsheet.CreateDefaultRibbonTabs(True)

			RemoveRibbonTab(spreadsheet, GetType(SRFileTab))
			RemoveRibbonTab(spreadsheet, GetType(SRInsertTab))
			RemoveRibbonTab(spreadsheet, GetType(SRPageLayoutTab))
			RemoveRibbonTab(spreadsheet, GetType(SRPageLayoutTab))
			RemoveRibbonTab(spreadsheet, GetType(SRFormulasTab))
			RemoveRibbonTab(spreadsheet, GetType(SRDataTab))
		End Sub

		Public Shared Sub HidePageLayoutTab(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet)
			RemoveRibbonTab(spreadsheet, GetType(SRPageLayoutTab))
		End Sub

		Public Shared Sub HideFormulasTab(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet)
			RemoveRibbonTab(spreadsheet, GetType(SRFormulasTab))
		End Sub

		Private Shared Sub HideRibbonItems(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet, ByVal itemTypes() As Type)
			ProcessRibbonItems(spreadsheet, itemTypes, False)
		End Sub
		Private Shared Sub ShowRibbonItemsOnly(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet, ByVal itemTypes() As Type)
			ProcessRibbonItems(spreadsheet, itemTypes, True)
		End Sub
		Private Shared Sub ProcessRibbonItems(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet, ByVal targetTypes() As Type, ByVal removeNotInList As Boolean)
			Dim groups = New List(Of RibbonGroup)()
			Dim items = New List(Of RibbonItemBase)()

			Dim tabs As List(Of RibbonTab) = spreadsheet.RibbonTabs.ToList()
			For Each tab As RibbonTab In tabs
				groups.AddRange(tab.Groups)
			Next tab
			For Each group As RibbonGroup In groups
				items.AddRange(group.Items)
			Next group

			RemoveItemsByTypes(items, targetTypes, removeNotInList)
			RemoveEmptyGroups(groups)
			RemoveEmptyTabs(spreadsheet, tabs)
		End Sub

		Private Shared Sub RemoveItemsByTypes(ByVal items As List(Of RibbonItemBase), ByVal targetTypes() As Type, ByVal removeNotInList As Boolean)
			Dim targetItem = New List(Of RibbonItemBase)()
			Dim itemsNotInList = New List(Of RibbonItemBase)()
			GroupItemsByType(items, targetTypes, targetItem, itemsNotInList)

			If removeNotInList Then
				RemoveItems(itemsNotInList)
			Else
				RemoveItems(targetItem)
			End If

		End Sub

		Private Shared Sub GroupItemsByType(ByVal items As List(Of RibbonItemBase), ByVal targetTypes() As Type, ByVal targetItem As List(Of RibbonItemBase), ByVal itemsNotInList As List(Of RibbonItemBase))
			For Each item As RibbonItemBase In items
				If targetTypes.Contains(item.GetType()) Then
					targetItem.Add(item)
				Else
					itemsNotInList.Add(item)
				End If
			Next item
		End Sub

		Private Shared Sub RemoveItems(ByVal items As List(Of RibbonItemBase))
			For Each item As RibbonItemBase In items
				item.Group.Items.Remove(item)
			Next item
		End Sub
		Private Shared Sub RemoveEmptyGroups(ByVal groups As List(Of RibbonGroup))
			For Each group As RibbonGroup In groups
				If group.Items.IsEmpty Then
					group.Tab.Groups.Remove(group)
				End If
			Next group
		End Sub
		Private Shared Sub RemoveEmptyTabs(ByVal spreadsheet As ASPxSpreadsheet.ASPxSpreadsheet, ByVal tabs As List(Of RibbonTab))
			For Each tab As RibbonTab In tabs
				If tab.Groups.IsEmpty Then
					spreadsheet.RibbonTabs.Remove(tab)
				End If
			Next tab
		End Sub
	End Class
End Namespace
